src/py/flwr/server/state/state_test.py

```diff
@@ -0,0 +1,421 @@
+# Copyright 2020 Adap GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests all state implemenations have to conform to."""
+# pylint: disable=no-self-use, invalid-name, disable=R0904
+
+import unittest
+from abc import abstractmethod
+from datetime import datetime, timezone
+from typing import List
+from uuid import uuid4
+
+from flwr.proto.node_pb2 import Node
+from flwr.proto.task_pb2 import Task, TaskIns, TaskRes
+from flwr.proto.transport_pb2 import ClientMessage, ServerMessage
+from flwr.server.state.in_memory_state import InMemoryState
+from flwr.server.state.state import State
+
+
+class StateTest(unittest.TestCase):
+    """Test all state implementations."""
+
+    # This is to true in each child class
+    __test__ = False
+
+    @abstractmethod
+    def state_factory(self) -> State:
+        """Provide state implementation to test."""
+        raise NotImplementedError()
+
+    def test_get_task_ins_empty(self) -> None:
+        """Validate that a new state has no TaskIns."""
+
+        # Prepare
+        state = self.state_factory()
+
+        # Execute
+        task_ins_list = state.get_task_ins(node_id=1, limit=10)
+
+        # Assert
+        assert len(task_ins_list) == 0
+
+    def test_get_task_ins_anonymous(self) -> None:
+        """Validate that a new state has no TaskIns."""
+
+        # Prepare
+        state = self.state_factory()
+        task_ins = create_task_ins(consumer_node_id=0, anonymous=True)
+        task_id = state.store_task_ins(task_ins=task_ins)
+
+        # Execute
+        task_ins_list = state.get_task_ins(
+            node_id=None,
+            limit=10,
+        )
+
+        # Assert
+        assert len(task_ins_list) == 1
+        assert task_ins_list[0].task_id == str(task_id)
+
+    def test_get_task_res_empty(self) -> None:
+        """Validate that a new state has no TaskRes."""
+
+        # Prepare
+        state = self.state_factory()
+
+        # Execute
+        task_res_list = state.get_task_res(task_ids={uuid4()}, limit=10)
```

Comment:
Wouldn't it be better to use `num_task_res` here?

Reply:
Good idea.